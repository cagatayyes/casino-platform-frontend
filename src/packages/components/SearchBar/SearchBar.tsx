import { CategoryType, SEARCH_GAMES_ENDPOINT } from 'casino/constants';
import { IGameData } from 'casino/types';
import { useEffect, useState } from 'react';
import SearchResults from 'casino/components/SearchResults/SearchResults';
import NoDataMessage from 'casino/components/NoDataMessage/NoDataMessage';
import Search from 'casino/components/Search/Search';
import axios, { AxiosResponse } from 'axios';
import { useParams } from 'react-router-dom';
import { CommonLoader } from 'casino/components/shared/CommonLoader';

interface ISearchBarProps {
    maxWidth: string;
    isSearch: boolean;
    setIsSearch:(isSearch:boolean) => void;
	showgameinfomodal:(gameId: number, gameLink: string) => void;
	handlegameload: (gameLink: string, gameName: string, gamelaunchpoint: string, categoryName?: string ) => void;
}

export const SearchBar = ({ maxWidth, isSearch, setIsSearch, showgameinfomodal, handlegameload }: ISearchBarProps): JSX.Element => {
	const [searchText, setSearchText] = useState<string>('');
	const [searchData, setSearchData] = useState<IGameData[] | null>([] as IGameData[]);
	const [isLoading, setIsLoading] = useState<boolean>(false);
	const [error, setError] = useState<string>('');
	const { categoryId: paramCategoryId, categorySlug } = useParams();
	const categoryId = paramCategoryId ? parseInt(paramCategoryId) : null;
	const categoryName = categorySlug ?? null;

	const getSearchResult = async (): Promise<void> => {
		try {
			const queryParam = categoryId != null && categoryName !== CategoryType.recentlyPlayed ?
				`searchText=${searchText}&categoryId=${categoryId}` : `searchText=${searchText}`;
			const apiURL = `${process.env.REACT_APP_API_URL}/${SEARCH_GAMES_ENDPOINT}?${queryParam}`;
			const response: AxiosResponse = await axios(apiURL);
			if (response.data.success) {
				setSearchData(response.data.data);
				setIsLoading(false);
				if (response?.data?.data?.length === 0) {
					setError('No search results found');
				}
			} else {
				setError('Oops! Something went wrong');
				setIsLoading(false);
			}

		} catch (error) {
			setError('Oops! Something went wrong');
			setIsLoading(false);
		}
	};

	useEffect(() => {
		if (isSearch) {
			const delayDebounce = setTimeout(() => {
				if (searchText.length >= 3) {
					getSearchResult();
				}
			}, 1000);

			return () => clearTimeout(delayDebounce);
		}
	}, [searchText]);

	const handleGameSearch = async (searchTerm: string): Promise<void> => {
		setIsLoading(true);
		if (searchTerm.length >= 3) {
			setSearchText(searchTerm);
			setIsSearch(true);
		} else {
			setSearchText(searchTerm);
			setIsSearch(false);
		}
	};

	const renderSearchResults = (): JSX.Element => {
		if (isSearch && searchData?.length) {
			return <SearchResults
				searchData={searchData}
				showgameinfomodal={showgameinfomodal}
				handlegameload={handlegameload}
			/>;
		} else {
			return <NoDataMessage
				message={error}
				maxWidth={maxWidth}
			/>;
		}
	};

	return (
		<>
			<div className='flex w-full justify-center'>
				<Search
					placeholder='Search games here etc...'
					className='flex w-full h-[32px] rounded-[4px] border-[1px] border-[#CBD4E01A] my-[14px] mx-[14px]'
					inputClassName='text-[#6C7086] text-[11px] font-normal pl-[5px] leading-[12.5px]'
					searchIconColor='#BBBFD7'
					clearIconColor='#9D9D9D'
					onChange={handleGameSearch}
				/>
			</div>
			{isSearch && !isLoading && renderSearchResults()}
			{isSearch && isLoading && <CommonLoader maxWidth={maxWidth} /> }
		</>

	);
};
