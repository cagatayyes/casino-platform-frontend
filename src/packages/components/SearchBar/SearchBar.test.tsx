/* eslint-disable react/prop-types */
import { render } from '@testing-library/react';
import { SearchBar } from './SearchBar';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

jest.mock('swiper/react', () => ({
	Swiper: ({ children }): JSX.Element => <div>{children}</div>,
	SwiperSlide: ({ children }): JSX.Element => <div>{children}</div>
}));

jest.mock('swiper/modules', () => ({
	Swiper: ({ children }): JSX.Element => <div>{children}</div>,
	SwiperSlide: ({ children }): JSX.Element => <div>{children}</div>
}));

describe('Search Bar', () => {
	it('should render Search Input and controllers', () => {
		const component = render(
			<BrowserRouter>
				<SearchBar
					maxWidth='640px'
					isSearch={true}
					setIsSearch={jest.fn()}
					showGameInfoModal={jest.fn()}
					handleGameLoad={jest.fn()}
				/>
			</BrowserRouter>
		);
		expect(component).toMatchSnapshot();
	});
});
