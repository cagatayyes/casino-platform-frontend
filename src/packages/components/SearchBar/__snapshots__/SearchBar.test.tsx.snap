// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar should render Search Input and controllers 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full justify-center"
      >
        <div
          class="inline-flex relative flex-row p-[10px] items-center  flex w-full h-[32px] rounded-[4px] border-[1px] border-[#CBD4E01A] my-[14px] mx-[14px]"
        >
          <div
            class="text-[#008751] mr-[5px] flex-grow-0 flex items-center"
          >
            <span
              class="w-[16px] h-[16px] material-symbols-outlined mt-[3px]"
              style="color: rgb(187, 191, 215); font-size: 16px;"
            >
              search
            </span>
          </div>
          <div
            class="relative flex flex-1 items-center"
          >
            <input
              class="bg-[transparent] outline-none w-full h-full text-[11px] pr-[24px] text-[11px] text-[#6C7086] text-[#6C7086] text-[11px] font-normal pl-[5px] leading-[12.5px]"
              placeholder="Search games here etc..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col items-center rounded-[10px] py-[17px] px-[14px] bg-[#0D1335] text-[12px]"
        style="max-width: 640px;"
      >
        <span
          class="flex w-full text-[#ffffff] justify-center"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full justify-center"
    >
      <div
        class="inline-flex relative flex-row p-[10px] items-center  flex w-full h-[32px] rounded-[4px] border-[1px] border-[#CBD4E01A] my-[14px] mx-[14px]"
      >
        <div
          class="text-[#008751] mr-[5px] flex-grow-0 flex items-center"
        >
          <span
            class="w-[16px] h-[16px] material-symbols-outlined mt-[3px]"
            style="color: rgb(187, 191, 215); font-size: 16px;"
          >
            search
          </span>
        </div>
        <div
          class="relative flex flex-1 items-center"
        >
          <input
            class="bg-[transparent] outline-none w-full h-full text-[11px] pr-[24px] text-[11px] text-[#6C7086] text-[#6C7086] text-[11px] font-normal pl-[5px] leading-[12.5px]"
            placeholder="Search games here etc..."
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="flex flex-col items-center rounded-[10px] py-[17px] px-[14px] bg-[#0D1335] text-[12px]"
      style="max-width: 640px;"
    >
      <span
        class="flex w-full text-[#ffffff] justify-center"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
